pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,429 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Do not make do anything if the cluster is not a GCP cluster.
+	val, ok := cd.Labels[clusterPlatformLabel]
+	if !ok || val != clusterPlatformGCP {
+		reqLogger.Info("not a gcp cluster")
+		return reconcile.Result{}, nil
+	}
+
+	// Do not do anything if the cluster is not a Red Hat managed cluster.
+	val, ok = cd.Labels[clusterDeploymentManagedLabel]
+	if !ok || val != "true" {
+		reqLogger.Info("not a managed cluster")
+		return reconcile.Result{}, nil
+	}
+
+	//Do not reconcile if cluster is installed or remove cleanup and remove project
+	if cd.Spec.Installed {
+		// TODO(Raf) Cleanup and remove project if being deleted once Hive is finished uninstalling
+		reqLogger.Info(fmt.Sprintf("cluster %v is in installed state", cd.Name))
+		return reconcile.Result{}, nil
+	}
+
+	// Get org creds from secret
+	creds, err := GetOrgGCPCreds(r.client, operatorNamespace)
+	if err != nil {
+		reqLogger.Error(err, "Could not get org Creds from secret", "Secret Name", orgGcpSecretName, "Operator Namespace", operatorNamespace)
+		return reconcile.Result{}, err
+	}
+
+	// Check if gcpSecretName in cd.Namesapce exists we are done
+	if SecretExists(r.client, gcpSecretName, cd.Namespace){
+		reqLogger.Info(fmt.Sprintf("Secret: %s already exists in Namespace: %s :: Nothing to do", gcpSecretName, cd.Namespace))
+		return reconcile.Result{}, nil
+	}
+
+	// Skip code block to create project for now until we have permissions to test
+	if false {
+		// Check if platform projectID string exists.
+		if cd.Spec.Platform.GCP.ProjectID != "" && cd.Spec.Platform.GCP.Region != "" {
+			// check that region is supported
+			if !supportedRegions[cd.Spec.Platform.GCP.Region] {
+				reqLogger.Error(ErrRegionNotSupported, "Regions is not supported", "Region", cd.Spec.Platform.GCP.Region)
+				// TODO(Raf) Should we be requeuing here or stop looping at this error
+				return reconcile.Result{}, ErrRegionNotSupported
+			}
+			// Get gcpclient with creds
+			gClient, err := gcpclient.NewClient(cd.Spec.GCP.ProjectID, creds)
+			if err != nil {
+				reqLogger.Error(err, "Could not get gcp client with secret creds", "Secret Name", orgGcpSecretName, "Operator Namespace", operatorNamespace)
+				return reconcile.Result{}, err
+			}
+
+			// TODO(Raf) Check that operation is complete before continuing , make sure project Name does not exits , How to handle those errors
+			_, err = gClient.CreateProject(orgParentFolderID)
+			if err != nil {
+				reqLogger.Error(err, "Could create project", "Parent Folder ID", orgParentFolderID, "Requested Project Name", cd.Spec.Platform.GCP.ProjectID, "Requested Region Name", cd.Spec.GCP.Region)
+				return reconcile.Result{}, err
+			}
+
+			// TODO(Raf) Set quotas
+			// TODO(Raf) Enable APIs
+		} else {
+			// TODO(Raf) Should we requeue in this case or not ?
```

Comment:
We need to validate this earlier. Potentially introduce validation function just after we got `CD` object from api server. Fail early.

Reply:
This why we dont have if `if else` too